contributors/design-proposals/storage/pv-monitoring-proposal.md

```diff
@@ -0,0 +1,128 @@
+
+#### For in-tree volume plugins(except local storage):
+
+We can add a new volume plugin interface: StatusCheckingVolumePlugin.
+```
+  type StatusCheckingVolumePlugin interface {
+       VolumePlugin
+
+      CheckStatus(spec *Spec) (string, error)
+  }
+```
+And each volume plugin will implement it. The entire monitoring controller workflow is:
+
+* Fill PV cache with initial data from etcd
+* Resync and check volumes status periodically
+* Taint PV if the volume status is abnormal
+* Remove the taints if the volume becomes normal again
+
+#### For out-tree volume plugins(except local storage):
+
+We can create a new controller called MonitorController like ProvisionController, 
+which is responsible for creating informers, watching Node and PV events and calling each plugin’s monitor functions. 
+And each volume plugin will create its own monitor to check its volumes’ status.
+
+#### For local storage:
+
+The local storage PV monitor consists of two parts
+
+* create a daemonset on every node, which is responsible for monitoring local PVs in that specific node, no matter the PVs are created manually or by provisioner;
+* create a monitor controller, which is responsible for watching PVs and Nodes events. PVs may be updated if they are unhealthy and we also need to react to node failure event.
+
+At the first phase, we can support local storage monitoring first.
+
+Take local storage as an example, detailed monitoring implementation may be like this:
+
+```
+ func (mon *minitor) CheckStatus(spec *v1.PersistentVolumeSpec) (string, error) {
+        // check if PV is local storage
+        if pv.Spec.Local == nil { 
+                glog.Infof("PV: %s is not local storage", pv.Name)
+                return ...
+        }
+        
+        // check if PV belongs to this node
+        if checkNodeAffinity(...) {
+                glog.Infof("PV: %s does not belong to this node", pv.Name)
+                return ...
+        }
+
+        // check volume health condition depending on device type
+        ...
+
+}
+```
+If monitor finds that one PV is unhealthy, it will mark the PV by adding annotations including timestamp. 
+The reaction controller then can react to this PV depending on the annotations and timestamp.
+
+### Reaction controller:
+
+Reaction controller will react to the PV update event (tainting PV by monitoring controller). 
+Different kinds of apps should have different reactions.
+
+At the first phase, we just consider the statefulset reaction.
+
+statefulset reaction: check the annotation timestamp, if the PV can recover within the predefined time interval, we will do nothing, 
+otherwise we need to delete the PVC bound to the unhealthy volume(PV) as well as pods referencing it.
+
+Reaction controller’s workflow is:
+
+* Fill PV cache from etcd;
+* Watch for PV update events;
+* Resync and populate periodically;
+* Delete related PVC and pods if needed (just for statefulsets and reclaim PV depending on reclaim policy);
```

Comment:
BTW, what application are you planning to use with this?  It may be good to mention in the beginning use cases what kinds of applications this is suitable for.

Reply:
if we want to delete the PV directly, the application should have data backup and can restore it or can tolerate data loss. btw: the PV protection feature may help